pageserver/src/tenant/timeline.rs

```diff
@@ -3927,6 +3924,199 @@ impl Timeline {
     }
 }
 
+#[derive(Clone)]
+/// Type representing a query in the ([`Lsn`], [`Key`]) space.
+/// In other words, a set of segments in a 2D space.
+///
+/// This representation has the advatange of avoiding hash map
+/// allocations for uniform queries.
+pub(crate) enum VersionedKeySpaceQuery {
+    /// Variant for queries at a single [`Lsn`]
+    Uniform { keyspace: KeySpace, lsn: Lsn },
+    /// Variant for queries at multiple [`Lsn`]s
+    Scattered {
+        keyspaces_at_lsn: Vec<(Lsn, KeySpace)>,
+
+        cached_total_keyspace: once_cell::sync::OnceCell<KeySpace>,
+        cached_high_watermark_lsn: once_cell::sync::OnceCell<Lsn>,
+    },
+}
+
+impl VersionedKeySpaceQuery {
+    pub(crate) fn uniform(keyspace: KeySpace, lsn: Lsn) -> Self {
+        Self::Uniform { keyspace, lsn }
+    }
+
+    pub(crate) fn scattered(keyspaces_at_lsn: Vec<(Lsn, KeySpace)>) -> Self {
+        Self::Scattered {
+            keyspaces_at_lsn,
+            cached_total_keyspace: Default::default(),
+            cached_high_watermark_lsn: Default::default(),
+        }
+    }
+
+    /// Returns the most recent (largest) LSN included in the query.
+    /// If any of the LSNs included in the query are invalid, returns
+    /// an error instead.
+    fn high_watermark_lsn(&self) -> Result<Lsn, GetVectoredError> {
+        match self {
+            Self::Uniform { lsn, .. } => {
+                if !lsn.is_valid() {
+                    return Err(GetVectoredError::InvalidLsn(*lsn));
+                }
+
+                Ok(*lsn)
+            }
+            Self::Scattered {
+                keyspaces_at_lsn,
+                cached_high_watermark_lsn,
+                ..
+            } => match cached_high_watermark_lsn.get() {
+                Some(cached) => Ok(*cached),
+                None => {
+                    let mut max_lsn = None;
+                    for (lsn, _keyspace) in keyspaces_at_lsn.iter() {
+                        if !lsn.is_valid() {
+                            return Err(GetVectoredError::InvalidLsn(*lsn));
+                        }
+                        max_lsn = std::cmp::max(max_lsn, Some(lsn));
+                    }
+
+                    if let Some(computed) = max_lsn {
+                        cached_high_watermark_lsn
+                            .set(*computed)
+                            .expect("cache invalidation");
+                        Ok(*computed)
+                    } else {
+                        Err(GetVectoredError::Other(anyhow!("empty input")))
+                    }
+                }
+            },
+        }
+    }
+
+    /// Returns the total keyspace being queried: the result of projecting
+    /// everything in the key dimensions onto the key axis.
+    fn total_keyspace(&self) -> KeySpace {
```

Comment:
At least one callsite can work with a borrow (&KeySpace) return type, I think. Let the callers do the cloning.

Reply:
We don't do caching anymore, so it makes sense to return the value.